src/tools/miri/tests/native-lib/pass/ptr_write_access.rs

```diff
@@ -0,0 +1,122 @@
+// Only works on Unix targets
+//@ignore-target: windows wasm
+//@only-on-host
+
+#![feature(box_as_ptr)]
+
+use std::mem::MaybeUninit;
+
+fn main() {
+    test_modify_int();
+
+    test_init_int();
+
+    test_init_array();
+
+    test_init_interior_mutable();
+
+    test_swap_ptr();
+
+    test_interact_dangling();
+
+    test_inner_alloc_exposed();
+}
+
+fn test_modify_int() {
+    extern "C" {
+        fn modify_int(ptr: *mut i32);
+    }
+
+    let mut x = 11;
+    unsafe { modify_int(&mut x) };
+
+    assert_eq!(x, 12);
+}
+
+fn test_init_int() {
+    extern "C" {
+        fn init_int(ptr: *mut i32);
+    }
+
+    let mut x = MaybeUninit::<i32>::uninit();
+    let x = unsafe {
+        init_int(x.as_mut_ptr());
+        x.assume_init()
+    };
+
+    assert_eq!(x, 21);
+}
+
+fn test_init_array() {
+    extern "C" {
+        fn init_array(ptr: *mut i32, len: usize);
+    }
+
+    const LEN: usize = 3;
+
+    let mut array = MaybeUninit::<[i32; LEN]>::uninit();
+    let array = unsafe {
+        init_array(array.as_mut_ptr().cast::<i32>(), LEN);
+        array.assume_init()
+    };
+
+    assert_eq!(array, [31; LEN]);
+}
+
+fn test_init_interior_mutable() {
```

Comment:
There is no `UnsafeCell` in this test, so what does it have to do with interior mutability?

I don't quite understand what this is testing. If this is intended to be the test involving a pointer in immutable memory pointing to mutable memory, then the test is wrong -- you need a `static IMM_PTR_TO_MUT: UnsafeInterior` for that.

Reply:
I called it that because of the pattern of allowing interior mutation underneath a `const`/shared pointer.

I should've left in a TODO to discuss with you:
This was indeed the test we mentioned but I kept running into issues, initially:

<details>
<summary>1. Idea with sharing static *mut i32.</summary>

```rust
fn test_init_static_inner() {
    extern "C" {
        fn init_static_inner(pptr: *const *mut i32);
    }

    static mut INNER: MaybeUninit<i32> = MaybeUninit::uninit();
    static STATIC: *mut i32 = unsafe { INNER.as_mut_ptr() };
    unsafe { init_static_inner(&STATIC) }

    assert_eq!(unsafe { INNER.assume_init() }, 61);
}
```
```rust
+error[E0277]: `*mut i32` cannot be shared between threads safely
+  --> tests/native-lib/pass/ptr_write_access.rs:LL:CC
+   |
+LL |     static STATIC: *mut i32 = unsafe { INNER.as_mut_ptr() };
+   |                    ^^^^^^^^ `*mut i32` cannot be shared between threads safely
+   |
+   = help: the trait `std::marker::Sync` is not implemented for `*mut i32`
+   = note: shared static variables must have a type that implements `Sync`
+
+error: aborting due to 1 previous error
```

</details>


<details>
<summary>2. Idea of sharing static *mut i32 hidden inside struct.
</summary>

```rust
fn test_init_static_inner() {
    extern "C" {
        fn init_static_inner(pptr: *const UnsafeInterior);
    }

    #[repr(C)]
    struct UnsafeInterior {
        ptr: *mut i32
    }
    unsafe impl Sync for UnsafeInterior {}

    static mut INNER: MaybeUninit<i32> = MaybeUninit::uninit();
    static STATIC: UnsafeInterior = UnsafeInterior { ptr: unsafe { INNER }.as_mut_ptr() };
    unsafe { init_static_inner(&STATIC) };

    assert_eq!(unsafe { INNER.assume_init() }, 51);
}
```
```rust
+error: encountered dangling pointer in final value of static
+  --> tests/native-lib/pass/ptr_write_access.rs:LL:CC
+   |
+LL |     static STATIC: UnsafeInterior = UnsafeInterior { ptr: unsafe { INNER }.as_mut_ptr() };
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 1 previous error
```

</details>

<details>
<summary>3. Idea of just let bindings (current, incorrect test).</summary>

```rust
fn test_init_interior_mutable() {
    extern "C" {
        fn init_interior_mutable(pptr: *const UnsafeInterior);
    }

    #[repr(C)]
    struct UnsafeInterior {
        mut_ptr: *mut i32
    }
    unsafe impl Sync for UnsafeInterior {}
    
    let mut x = MaybeUninit::<i32>::uninit();
    let unsafe_interior = UnsafeInterior { mut_ptr: x.as_mut_ptr() };
    unsafe { init_interior_mutable(&unsafe_interior) };

    assert_eq!(unsafe { x.assume_init() }, 51);
}
```

</details>

Reply:
That fixes the test – thank you :-)
(though it uncovers a nasty `BorrowMutError`...)

Concerning my inaccurate terminology: I see, I'll rename it back to `test_init_static_inner` (and the struct to `UnsafeInner`).

Reply:
> (though it uncovers a nasty BorrowMutError...)

Where does that error trigger?

>  (and the struct to UnsafeInner).

Please rename it to `SyncPtr` or so; similar structs already exist in a bunch of places and that's usually how we call it.

Reply:
(Both fixed. :-)